### ISSUE TITLE
What's the idea behind asymmetric padding during downsampling?

### ISSUE THREAD

**Issue by Arksyd96** [Stability-AI/stablediffusion#355]
What's the idea behind this down-sampling with asymmetric padding :
Why don't we just use a symmetric padding of 1, everything would fit perfectly.
``
class Downsample(nn.Module):
    def __init__(self, in_channels, with_conv):
        super().__init__()
        self.with_conv = with_conv
        if self.with_conv:
            # no asymmetric padding in torch conv, must do it ourselves
            self.conv = torch.nn.Conv2d(in_channels,
                                        in_channels,
                                        kernel_size=3,
                                        stride=2,
                                        padding=0)

    def forward(self, x):
        if self.with_conv:
            pad = (0,1,0,1)
            x = torch.nn.functional.pad(x, pad, mode="constant", value=0)
            x = self.conv(x)
        else:
            x = torch.nn.functional.avg_pool2d(x, kernel_size=2, stride=2)
        return x
``

**Comment by Arksyd96** [comment]
No, unfortunately not yet !

**Comment by jaycee-tian** [comment]
> What's the idea behind this down-sampling with asymmetric padding : Why don't we just use a symmetric padding of 1, everything would fit perfectly.
> 
> ```
> class Downsample(nn.Module):
>     def __init__(self, in_channels, with_conv):
>         super().__init__()
>         self.with_conv = with_conv
>         if self.with_conv:
>             # no asymmetric padding in torch conv, must do it ourselves
>             self.conv = torch.nn.Conv2d(in_channels,
>                                         in_channels,
>                                         kernel_size=3,
>                                         stride=2,
>                                         padding=0)
> 
>     def forward(self, x):
>         if self.with_conv:
>             pad = (0,1,0,1)
>             x = torch.nn.functional.pad(x, pad, mode="constant", value=0)
>             x = self.conv(x)
>         else:
>             x = torch.nn.functional.avg_pool2d(x, kernel_size=2, stride=2)
>         return x
> ```

I also wonder why it is.

**Comment by GGYtilE** [comment]
Has anyone found the answer? I'm curious why.

**Comment by weiaicunzai** [comment]
I'm also curious about this question. Does anyone have an idea?

**Comment by Arksyd96** [comment]
Clearly, no one wants to answer to this question lol

**Comment by lxxfcp** [comment]
+1 I'm also curious

**Comment by nicolasch96** [comment]
To get a larger Box blur while keeping 0 shift with the average pooling version. Technically they reproduce what average pooling does but with a larger kernel same stride.
